[DOC] Supporting doc changes for additional CRD status fields (#1833)
strimzi/strimzi-kafka-operator: documentation/book/con-custom-resources-status.adoc

= {ProductName} custom resource status

The `status` property of a {ProductName} custom resource publishes information about the resource to users and tools that need it.

Several resources have a `status` property, as described in the following table.

[cols="3*",options="header",stripes="none",separator=¦]
|===

m¦{ProductName} resource
¦Schema reference
¦Publishes status information on...

m¦Kafka
¦xref:type-KafkaStatus-reference[]
¦The Kafka cluster.

m¦KafkaConnect
¦xref:type-KafkaConnectStatus-reference[]
¦The Kafka Connect cluster, if deployed.

m¦KafkaConnectS2I
¦xref:type-KafkaConnectS2Istatus-reference[]
¦The Kafka Connect cluster with Source-to-Image support, if deployed.

m¦KafkaMirrorMaker
¦xref:type-KafkaMirrorMakerStatus-reference[]
¦The Kafka MirrorMaker tool, if deployed.

m¦KafkaUser
¦xref:type-KafkaUserStatus-reference[]
¦Kafka users in your cluster.

m¦KafkaBridge
¦xref:type-KafkaBridgeStatus-reference[]
¦The {ProductName} Kafka Bridge, if deployed.

|===

> //When dev work is complete, add `KafkaTopicStatus` to the above table. `KafkaTopicStatus`: Kafka topics and their partitions

Reviewer: should you remove this comment from the doc?

Author: Yes, I'll remove it.

The `status` property of a resource provides information on the resource's:

* _Current state_, in the `status.conditions` property

* _Last observed generation_, in the `status.observedGeneration` property

The `status` property also provides resource-specific information. For example:

* `KafkaConnectStatus` provides the REST API endpoint for Kafka Connect connectors.

* `KafkaUserStatus` provides the user name of the Kafka user and the `Secret` in which their credentials are stored.

> * `KafkaBridgeStatus` provides the HTTP address at which external client applications can access the Bridge service.

Reviewer: maybe we should say "HTTP based URL" based on the naming we are using in the implementation. Wdyt @tombentley ?

Author: Or:
"provides the URL at which external client applications can access the Bridge service"

A resource's _current state_ is useful for tracking progress related to the resource achieving its _desired state_, as defined by the `spec` property. The status conditions provide the time and reason the state of the resource changed and details of events preventing or delaying the operator from realizing the resource's desired state.

> The _last observed generation_ is the generation of the resource that was most recently observed by the operator. If the value of `observedGeneration` is less than the value of `metadata.generation`, the operator has not yet processed the latest update to the resource. If these values are the same, the status information reflects the most recent changes to the resource.

Reviewer 2: ```suggestion
The _last observed generation_ is the generation of the resource that was most recently observed by the operator. If the value of `observedGeneration` is different from the value of `metadata.generation`, the operator has not yet processed the latest update to the resource. If these values are the same, the status information reflects the most recent changes to the resource.
```

= {ProductName} custom resource status

The `status` property of a {ProductName} custom resource publishes useful information about the resource to users and tools that need it.

Several resources have a `status` property, as described in the following table.

[cols="2*",options="header",stripes="none",separator=¦]
|===

¦{ProductName} Resource
¦Publishes information on...

m¦xref:type-KafkaStatus-reference[]
¦The Kafka cluster.

> m¦xref:type-KafkaTopicStatus-reference[]

Reviewer 3: Does this link already exist? If not we will need to wait until it is created by @tombentley before merging this for the CI pipelines.

Reviewer 2: It might be better to remove/comment this and merge this PR now, than wait for a PR with `KafkaTopic.status`.

Author: I'll remove it and add a comment to the module.

¦Kafka topics and their partitions.

m¦xref:type-KafkaConnectStatus-reference[]
¦The Kafka Connect cluster, if deployed.

m¦xref:type-KafkaConnectS2Istatus-reference[]
¦The Kafka Connect cluster with Source-to-Image support, if deployed.

m¦xref:type-KafkaMirrorMakerStatus-reference[]
¦The Kafka MirrorMaker tool, if deployed.

m¦xref:type-KafkaUserStatus-reference[]
¦Kafka users.

m¦xref:type-KafkaBridgeStatus-reference[]
¦The {ProductName} Kafka Bridge, if deployed.

|===

The `status` property includes information on:

* The _current state_ of the resource described in _status conditions_

> * The resource's last `observedGeneration` (the generation of the resource to which the status applies)

Reviewer 3: Should we explain a bit more how does this related to the `generation` field in the `metadata` section?

Author: I can't find the any docs on the `generation` field in the API Reference. Is it missing?

Reviewer 3: It is a Kubernetes field, so it is not in our API reference. Basically, the `metadata.generation` field is an integer which is incremented everytime the resource is modified (i.e. the user changes it with `kubectl apply`, `kubectl edit` etc.). The `status.observedGeneration` basically mirrors the `metadata.generation` field to show what was the last generation of this resource which the operator has already seen.

The usefulness for the user is that the user can compare these number and when they are the same it means that the operator saw the latest version and the status is valid for it. If the `observedGeneration` is smaller then the `generation`, it means that the operator did not process the last update to the custom resource yet.

I didn't find any link in Kubernetes docs either - apart from the API reference which doesn't explain much. So I think it would be good to add a bit from what I described here into the docs.

Author: Thanks for this. I've added a brief description of "last observed generation".

* Other information specific to the resource, for example the REST API endpoint for Kafka Connect connectors.

> A resource's current state is useful for tracking progress related to the resource achieving its _desired state_, as defined by the `spec` property. The status conditions provide the time and reason the state of the resource changed and details of events preventing or delaying the Operator from realizing the resource's desired state.

Reviewer 4: @laidan6000 should operator be capitalized? It typically isn't in other documentation I've seen.

Author: Good spot. We're not capitalizing operator. I'll change to lower-case.

----
<1> Status `conditions` describe criteria related to the status that cannot be deduced from the existing resource information, or are specific to the instance of a resource.
<2> The `Ready` condition indicates whether the Cluster Operator currently considers the Kafka cluster able to handle traffic.

> <3> The `observedGeneration` indicates the generation of the `Kafka` custom resource that was last reconciled by the Cluster Operator.

Reviewer 4: @laidan6000 is Cluster Operator a proper noun? Wondering about capitalizing operator.

Author: Yes, the current Streams docs use "Cluster Operator" as a proper noun throughout.
